Refresh FFM documents (#914)
openjdk/panama-foreign: doc/panama_ffi.md

```java
MemoryLayout SIZE_T = Linker.nativeLinker().canonicalLayouts().get("size_t");
```

Composite types are modeled as group layouts. More specifically, a C struct type maps to a `StructLayout`, whereas a C `union` type maps to a `UnionLayout`. When defining a struct or union layout, clients must pay attention to the size and alignment constraint of the corresponding composite type definition in C. For instance, padding between two struct fields must be modeled explicitly, by adding an adequately sized padding layout member to the resulting struct layout.

Finally, pointer types such as `int**`, and `int(*)(size_t*, size_t*)` are modeled as address layouts. When the spatial bounds of the pointer type are known statically, the address layout can be associated with a *target layout*. For instance, a pointer that is known to point to a C `int[2]` array can be modelled as follows:

```java
ValueLayout.ADDRESS.withTargetLayout(
        MemoryLayout.sequenceLayout(2,
                                    Linker.nativeLinker().canonicalLayouts().get("int")));
```

> For a more exhaustive examples of mapping between C types and layouts, please refer to the [appendix](#c-types-mapping-in-linuxx64). In the following sections, we will assume Linux/x64 as our target platform.

Reviewer: Misplaced 's'?

```suggestion
For a more exhaustive example of mappings between C types and layouts, please refer to the [appendix](#c-types-mapping-in-linuxx64). In the following sections, we will assume Linux/x64 as our target platform.
```

                                    Linker.nativeLinker().canonicalLayouts().get("int")));
```

> For a more exhaustive examples of mappings between C types and layouts, please refer to the [appendix](#c-types-mapping-in-linuxx64). In the following sections, we will assume Linux/x64 as our target platform.

Reviewer: ```suggestion
For more exhaustive examples of mappings between C types and layouts, please refer to the [appendix](#c-types-mapping-in-linuxx64). In the following sections, we will assume Linux/x64 as our target platform.
```

> 

Reviewer 2: ```suggestion
```

```java
try (Arena arena = Arena.ofConfined()) {
    MemorySegment comparFunc = linker.upcallStub(comparHandle, comparDesc, arena);

>     MemorySegment array = session.allocateFrom(0, 9, 3, 4, 6, 5, 1, 8, 2, 7);

Reviewer 3: I think it should read `arena.allocateFrom(...)`

```java
try (Arena arena = Arena.ofConfined()) {
    int res = (int)printf.invokeExact(arena.allocateFrom("%d plus %d equals %d"), 2, 2, 4); //prints "2 plus 2 equals 4"
}
```

> While this works, and provides optimal performance, it has some limitations<a href="#2"><sup>3</sup></a>:

Reviewer 2: ```suggestion
While this works, and provides optimal performance, it has some limitations<a href="#3"><sup>3</sup></a>:
```

}
```

* <a id="1"/>(<sup>1</sup>):<small> In reality this is not entirely new; even in JNI, when you call a `native` method the VM trusts that the corresponding implementing function in C will feature compatible parameter types and return values; if not a crash might occur.</small>
* <a id="2"/>(<sup>2</sup>):<small> Linker options can be used to customize the linkage request in various ways, for instance to allow clients to pass heap segments to native functions without copying, to remove Java to native thread transitions and to save the state of special runtime variables (such as `errno`). </small>

> * <a id="3"/>(<sup>2</sup>):<small> Previous iterations of the FFM API provided a `VaList` class that could be used to model a C `va_list`. This class was later dropped from the FFM API as too implementation specific. It is possible that a future version of the `jextract` tool might provide higher-level bindings for variadic calls. </small>

Reviewer 2: ```suggestion
* <a id="3"/>(<sup>3</sup>):<small> Previous iterations of the FFM API provided a `VaList` class that could be used to model a C `va_list`. This class was later dropped from the FFM API as too implementation specific. It is possible that a future version of the `jextract` tool might provide higher-level bindings for variadic calls. </small>
```
